README.md
# scheduler
A simple tool that provides a easy way for writing and deploying cron jobs by C++

# Requirements

* [json](https://github.com/nlohmann/json)
* [boost](http://www.boost.org/)

# Getting Started

* git clone https://github.com/nanjingboy/scheduler ~/workspace/scheduler
* cd ~/workspace/scheduler
* make
* bin/scheduler init

#### Example schedule.json file

```javascript
[
  {
    "minutes": [0],
    "hours": [12],
    "daysOfTheMonth": [1, 2],
    "months": [1],
    "daysOfTheWeek": [6, 7],
    "command": "ENV=test HOME=/Users/tom balabalababala 2>> error.log"
  }
]
```
* minutes must between 1 and 59.
* hours must between 1 and 23.
* daysOfTheMonth must between 1 and 30.
* months must between 1 and 12.
* daysOfTheWeek must between 0 and 7.

* The value of minutes, hours, daysOfTheMonth, months, daysOfTheWeek can also be a string, [click to see the fromat](http://www.nncron.ru/help/EN/working/cron-format.htm)

# Usage

```shell
$ bin/scheduler

Usage: scheduler [command] [option]

Options:

  -p, --path  Path of the initial schedule file
  -u, --user  Specify the name of the user whose crontab is to be tweaked
  -h, --help  Display this help message
  -V, --version  Display this application version

Available Commands:

  init  Create an initial schedule file
  write  Write cron jobs that defined in initial schedule file to crontab
  clear  Clear cron jobs that defined in initial schedule file from crontab

```

# License

MIT

CHANGELOG.md
## 0.1.0 2017-08-20
* Initial release

src/main.cpp
#include <iostream>
#include "cli.h"
#include "commander.h"

int main(int argc, const char * argv[]) {
  Option path_option;
  path_option.short_name = "-p";
  path_option.name = "path";
  path_option.description = "Path of the initial schedule file";

  Option user_option;
  user_option.short_name = "-u";
  user_option.name = "user";
  user_option.description = "Specify the name of the user whose crontab is to be tweaked";

  Command init_command;
  init_command.name = "init";
  init_command.description = "Create an initial schedule file";
  init_command.callback = Cli::init;

  Command write_command;
  write_command.name = "write";
  write_command.description = "Write cron jobs that defined in initial schedule file to crontab";
  write_command.callback = Cli::write;

  Command clear_command;
  clear_command.name = "clear";
  clear_command.description = "Clear cron jobs that defined in initial schedule file from crontab";
  clear_command.callback = Cli::clear;

  Commander commander("scheduler", "0.1.0");
  commander.add_option(path_option);
  commander.add_option(user_option);
  commander.add_command(init_command);
  commander.add_command(write_command);
  commander.add_command(clear_command);
  commander.run(argc, argv);
  return 0;
}

includes/cli.h
#ifndef CLI_H_
#define CLI_H_

#include <map>
#include <vector>
#include <string>
#include <boost/filesystem.hpp>

class Cli {

private:
  static boost::filesystem::path parse_path(const std::map<std::string, std::string> & option_values);
  static std::string parse_user(const std::map<std::string, std::string> & option_values);
  static std::vector<std::string> parse_jobs_can_not_remove(const std::string & path, const std::string & user);
  static std::string parse_schedule_time(const std::vector<int> & times, int min_value, int max_value);
  static std::vector<std::string> parse_schedule_file(const std::string & path);
  static void write_crontab(const std::vector<std::string> & jobs, const std::string & user);

public:
  static void init(const std::map<std::string, std::string> & option_values);
  static void write(const std::map<std::string, std::string> & option_values);
  static void clear(const std::map<std::string, std::string> & option_values);
};

#endif //CLI_H_

src/commander.cpp
#include <iostream>
#include <cstdlib>
#include "commander.h"

using namespace std;

Commander::Commander(const string & name, const string & version)
  :m_name(name), m_version(version) {
}

int Commander::find_option(const string & name) {
  size_t option_count = m_options.size();
  for (size_t index = 0; index < option_count; index++) {
    Option option = m_options[index];
    if (name == option.short_name || name == "--" + option.name) {
      return index;
    }
  }
  return -1;
}

int Commander::find_command(const string & name) {
  size_t command_count = m_commands.size();
  for (size_t index = 0; index < command_count; index++) {
    if (m_commands[index].name == name) {
      return index;
    }
  }
  return -1;
}

void Commander::print_help() {
  cout << endl << "Usage: " << m_name << " [command] [option]" << endl << endl;
  cout << "Options:" << endl << endl;
  for (Option option: m_options) {
    cout << "  " << option.short_name << ", --" << option.name << "  " << option.description << endl;
  }

  cout << "  -h, --help  Display this help message" << endl;
  cout << "  -V, --version  Display this application version" << endl << endl;
  if (m_commands.size() > 0) {
    cout << "Available Commands:" << endl << endl;
    for (Command command: m_commands) {
      cout << "  " << command.name << "  " << command.description << endl;
    }
  }
}

void Commander::print_version() {
  cout << m_version << endl;
}

void Commander::print_option_error(const string & option) {
  cout << endl << "  error: unknown option '" << option << "'" << endl;
  exit(1);
}

void Commander::add_option(const struct Option & option) {
  m_options.push_back(option);
}

void Commander::add_command(const struct Command & command) {
  m_commands.push_back(command);
}

void Commander::run(int argc, const char * argv[]) {
  if (argc < 2) {
    print_help();
  } else {
    string first_arg = argv[1];
    if (first_arg.find("-") == 0) {
      if (first_arg == "-h" || first_arg == "--help") {
        print_help();
      } else if (first_arg == "-V" || first_arg == "--version") {
        print_version();
      } else {
        print_option_error(first_arg);
      }
    } else {
      int command_index = find_command(first_arg);
      if (command_index > -1) {
        map<string, string> option_values;
        if (argc >= 3) {
          for (int index = 2; index < argc; index++) {
            string arg = argv[index];
            if (arg == "-h" || arg == "--help") {
              print_help();
              return;
            }
            if (arg == "-V" || arg == "--version") {
              print_version();
              return;
            }
            if (arg.find("-") == 0) {
              int option_index = find_option(arg);
              if (option_index == -1) {
                print_option_error(arg);
              } else if (index < argc - 1) {
                string value = argv[index + 1];
                if (value.find("-") != 0) {
                  option_values[m_options[option_index].name] = value;
                }
              }
            }
          }
        }
        Command command = m_commands[command_index];
        if (command.callback) {
          command.callback(option_values);
        }
      }
    }
  }
}

includes/commander.h
#ifndef COMMANDER_H_
#define COMMANDER_H_

#include <map>
#include <vector>
#include <string>

typedef void (*command_callback)(const std::map<std::string, std::string> &);

struct Option {
  std::string short_name;
  std::string name;
  std::string description;
};

struct Command {
  std::string name;
  std::string description;
  command_callback callback;
};

class Commander {

private:
  std::string m_name;
  std::string m_version;
  std::vector<Option> m_options;
  std::vector<Command> m_commands;

  int find_option(const std::string & name);
  int find_command(const std::string & name);

  void print_help();
  void print_version();
  void print_option_error(const std::string & option);

public:
  Commander(const std::string & name, const std::string & version);
  void add_option(const struct Option & option);
  void add_command(const struct Command & command);
  void run(int argc, const char * argv[]);
};

#endif //COMMANDER_H_

src/cli.cpp
#include <iostream>
#include <cstdlib>
#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>
#include <boost/process.hpp>
#include <json.hpp>

#include "cli.h"

using namespace std;
using namespace boost;

filesystem::path Cli::parse_path(const map<string, string> & option_values) {
  if (option_values.find("path") == option_values.end()) {
    return (filesystem::current_path() /= "schedule.json");
  }
  return filesystem::absolute(option_values.at("path"));
}

string Cli::parse_user(const map<string, string> & option_values) {
  if (option_values.find("user") == option_values.end()) {
    passwd *pwd;
    pwd = getpwuid(geteuid());
    return pwd->pw_name;
  }
  return option_values.at("user");
}

vector<string> Cli::parse_jobs_can_not_remove(const string & path, const string & user) {
  vector<string> jobs;
  process::ipstream pipe_stream;
  process::system("sudo crontab -l -u " + user, process::std_out > pipe_stream, process::std_err > process::null);

  string job;
  bool removeable = false;
  while (pipe_stream && getline(pipe_stream, job) && !job.empty()) {
    string start_comment = "#BEGIN DEFINE CRON JOBS FROM:" + path;
    if (job == start_comment) {
      removeable = true;
      continue;
    }

    string end_comment = "#END DEFINE CRON JOBS FROM:" + path;
    if (job == end_comment) {
      removeable = false;
      continue;
    }
    if (!removeable) {
      jobs.push_back(job);
    }
  }
  return jobs;
}

string Cli::parse_schedule_time(const vector<int> & times, int min_value, int max_value) {
  string result = "";
  for (int time: times) {
    if (time >= min_value && time <= max_value) {
      result += to_string(time) + ",";
    }
  }
  if (result != "") {
    result.pop_back();
    return result;
  }

  return "*";
}

vector<string> Cli::parse_schedule_file(const string & path) {
  vector<string> jobs;
  ifstream ifs(path);
  if (ifs.is_open()) {
    nlohmann::json json;
    ifs >> json;
    ifs.close();
    if (json.is_array()) {
      if (json.size() > 0) {
        for (auto & element: json) {
          string command = "";
          if (element["command"].is_string()) {
            command = element["command"];
          }
          if (command == "") {
            continue;
          }
          string sections[] = { "*", "*", "*", "*", "*" };
          if (element["minutes"].is_string()) {
            sections[0] = element["minutes"];
          } else if (element["minutes"].is_array()) {
            sections[0] = parse_schedule_time(element["minutes"], 0, 59);
          }

          if (element["hours"].is_string()) {
            sections[1] = element["hours"];
          } else if (element["hours"].is_array()) {
            sections[1] = parse_schedule_time(element["hours"], 0, 23);
          }

          if (element["daysOfTheMonth"].is_string()) {
            sections[2] = element["daysOfTheMonth"];
          } else if (element["daysOfTheMonth"].is_array()) {
            sections[2] = parse_schedule_time(element["daysOfTheMonth"], 1, 31);
          }

          if (element["months"].is_string()) {
            sections[3] = element["months"];
          } else if (element["months"].is_array()) {
            sections[3] = parse_schedule_time(element["months"], 1, 12);
          }

          if (element["daysOfTheWeek"].is_string()) {
            sections[4] = element["daysOfTheWeek"];
          } else if (element["daysOfTheWeek"].is_array()) {
            sections[4] = parse_schedule_time(element["daysOfTheWeek"], 0, 7);
          }

          string job = "";
          for (const string & section: sections) {
            job += section + " ";
          }
          job += " " + command;
          jobs.push_back(job);
        }
      }
    } else {
      cout << "Schedule file format is error: " << path << endl;
      exit(1);
    }
  }
  return jobs;
}

void Cli::write_crontab(const vector<string> & jobs, const string & user) {
  if (jobs.size() == 0) {
    process::system("sudo crontab -r -u " + user, process::std_err > process::null);
    return;
  }
  filesystem::path crontab_tmp_path = filesystem::temp_directory_path() /= "crontab.tmp";
  ofstream ofs(crontab_tmp_path.string());
  if (ofs.is_open()) {
    for (const string & job: jobs) {
      ofs << job << endl;
    }
    ofs.close();
  }
  process::system("sudo crontab -u " + user + " " + crontab_tmp_path.string());
  filesystem::remove(crontab_tmp_path);
}

void Cli::init(const map<string, string> & option_values) {
  filesystem::path path = parse_path(option_values);
  string path_string = filesystem::weakly_canonical(path).string();
  if (filesystem::is_directory(path)) {
    cout << "Initial file can't be directory: " << path_string << endl;
    exit(1);
  }
  if (filesystem::exists(path)) {
    cout << "Initial file already exists: " << path_string << endl;
    exit(1);
  }
  filesystem::path parent_path = path.parent_path();
  if (!filesystem::is_directory(parent_path)) {
    cout << "Directory does not exist: " << filesystem::weakly_canonical(parent_path).string() << endl;
    exit(1);
  }

  ofstream ofs(path_string);
  if (ofs.is_open()) {
    ofs << "[" << endl << "]" << endl;
    ofs.close();
  }
  cout << "Initial file created: " << path_string << endl;
}

void Cli::write(const map<string, string> & option_values) {
  vector<string> schedule_file_jobs;
  filesystem::path path = parse_path(option_values);
  string path_string = filesystem::weakly_canonical(path).string();
  if (filesystem::exists(path)) {
    schedule_file_jobs = parse_schedule_file(path_string);
    if (schedule_file_jobs.size() > 0) {
      schedule_file_jobs.insert(schedule_file_jobs.begin(), "#BEGIN DEFINE CRON JOBS FROM:" + path_string);
      schedule_file_jobs.push_back("#END DEFINE CRON JOBS FROM:" + path_string);
    }
  }

  string user = parse_user(option_values);
  vector<string> jobs = parse_jobs_can_not_remove(path_string, user);
  for (const string & job: schedule_file_jobs) {
    jobs.push_back(job);
  }
  write_crontab(jobs, user);
  cout << "Crontab file has been written" << endl;
}

void Cli::clear(const map<string, string> & option_values) {
  string user = parse_user(option_values);
  string path = filesystem::weakly_canonical(parse_path(option_values)).string();
  write_crontab(parse_jobs_can_not_remove(path, user), user);
  cout << "Crontab file has been cleared" << endl;
}